docs/architecture/tls13-experimental.md

```diff
@@ -66,3 +66,245 @@ together with their level of testing:
   as part of `MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL`:
 
   - Reader ([`library/mps_reader.h`](../../library/mps_reader.h))
+
+
+MVP definition
+--------------
+
+The TLS 1.3 MVP implements only the client side of the protocol.
+The TLS 1.3 MVP does not support the handling of server HelloRetryRequest and
+CertificateRequest messages. If it receives one of those messages, it aborts
+the handshake with an handshake_failure closure alert.
+
+- Supported cipher suites: depends on the library configuration. Potentially
+  all of them:
+  TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256,
+  TLS_AES_128_CCM_SHA256 and TLS_AES_128_CCM_8_SHA256.
+
+- Supported ClientHello extensions:
+
+                                 MVP   Prototype
+                                       (for comparison)
+
+  server_name                    no    YES
+  max_fragment_length            no    YES
+  status_request                 no    no
+  supported_groups               YES   YES
+  signature_algorithms           YES   YES
+  use_srtp                       no    no
+  heartbeat                      no    no
+  apln                           no    YES
+  signed_certificate_timestamp   no    no
+  client_certificate_type        no    no
+  server_certificate_type        no    no
+  padding                        no    no
+  key_share                      YES   YES
```

Comment:
Btw, I just checked if the RFC defines any mandatory-to-implement groups, and [it does](https://datatracker.ietf.org/doc/html/rfc8446#section-9.1):

```
   In the absence of an application profile standard specifying
   otherwise:
   [...] A
   TLS-compliant application MUST support key exchange with secp256r1
   (NIST P-256) and SHOULD support key exchange with X25519 [RFC7748].
```
So, we might chose to keep the current approach of sending only one key share and not supporting HRR, but tweak the default order of curves for TLS 1.3 so that P-256 comes first.

Reply:
Thanks for the pointer, very useful.

Reply:
I've added a note about `key_share` extension.

Reply:
> tweak the default order of curves for TLS 1.3 so that P-256 comes first

Unfortunately, as @paul-elliott-arm noted, the default order of curves on the client cannot depend on the protocol version, so this would mean preferring P256 to 25519 for TLS 1.2 too.

> I strongly suspect that most TLS 1.3 servers will accept both of the curves

So how about keeping the current curve order, which has 25519 before p256? This would make it less-efficient/not-implemented-yet (I don't know which) to connect to a server that doesn't support curve25519, but since that's unlikely to happen, it would be ok for the MVP, wouldn't it?

I think that not affecting TLS 1.2 operation at all is a worthwhile goal for the 1.3 MVP.

Reply:
I think keeping that current order and hoping that the server will support x25519 as well is probably the best thing to do:
- servers are very likely to support x25519 even if it's only a SHOULD, as it's the most efficient curve out there and servers tend to care about that
- this is just about the defaults, users can still use `mbedtls_ssl_conf_curves()` to set another order if they want/need to
- as you say, it's desirable not to affect 1.2

However, I'd like to question the fact that the order cannot depend on the protocol version in the short term: until we implement version negotiation (which is not part of the MVP IIRC), I think we can, it would just be inconvenient / a hack. When we start the handshake, we know what protocol is going to be used, so we check if the list of curves is the default, and if it is, swap it for the 1.3 default if using 1.3. (This would be somewhat similar to what's been done in #4859 - could probably even re-use some of the machinery.) Then we'd have to implement HRR and get rid of that hack before we implement version negotiation.

That said, I don't think it would be worth the complexity, and the time it would take to implement this would be better spent getting started on support for HRR for example.